source/math/FloatVectorList.ooc

```diff
@@ -381,4 +381,88 @@ FloatVectorList: class extends VectorList<Float> {
 			result add(this[i] clamp(floor, ceiling))
 		result
 	}
+	findMaxIndex: func -> (Int, Float) {
+		result := Float negativeInfinity
+		index := 0
+		for (i in 0 .. this count)
+			if (result < this[i]) {
+				result = this[i]
+				index = i
+			}
+		(index, result)
+	}
+	crossCorrelation: func (other: This, vectorLength: Int) -> Float {
+		upperPart := 0.0f
+		lowerPart := 0.0f
+		lowerPart2 := 0.0f
+		for (i in 0 .. vectorLength) {
+			upperPart += ((this[i] - this mean) * (other[i] - other mean))
+			lowerPart += ((this[i] - this mean) * (this[i] - this mean))
+			lowerPart2 += ((other[i] - other mean) * (other[i] - other mean))
+		}
+		upperPart / sqrt(lowerPart * lowerPart2)
+	}
+	shift: func (offset: Int, filledValue := 0.0f) -> This {
+		result := This new(this count)
+		if (offset < 0) {
+			for (i in 0 .. this count) {
+				if ((i - offset) <= this count)
+					result add(this[i - offset])
+				else
+					result add(filledValue)
+			}
+		}
+		else {
+			for (i in 0 .. this count) {
+				if (i < offset)
+					result add(filledValue)
+				else
+					result add(this[i - offset])
+			}
+		}
+		result
+	}
+	// offset < 0 means it shifts to left
+	findCrossCorrelationOffset: func (other: This, offsetStart, offsetEnd: Int) -> CorrelationData {
+		println("There are " + this count toString() + " points, please wait for calculating...")
```

Comment:
This part takes long time if there are too many points, I thought it's better to let user know it's still running well.